Libraries/rocSPARSE/preconditioner/csritilu0/main.cpp

```diff
@@ -0,0 +1,310 @@
+    //     (  6   13  20  26  32  38  )
+    //     (  8   18  29  40  50  60  )
+    //     (  10  23  38  54  70  85  )
+    //     (  12  28  47  68  90  112 )
+
+    // Number of rows and columns of the input matrix.
+    constexpr rocsparse_int n = 6;
+
+    // Number of non-zero elements.
+    constexpr rocsparse_int nnz = 36;
+
+    // CSR values vector.
+    // clang-format off
+    constexpr std::array<double, nnz> h_csr_val
+        = {2,   3,   4,   5,   6,   7,
+           4,   8,   11,  14,  17,  20,
+           6,   13,  20,  26,  32,  38,
+           8,   18,  29,  40,  50,  60,
+           10,  23,  38,  54,  70,  85,
+           12,  28,  47,  68,  90,  112};
+    // clang-format on
+
+    // CSR row pointers vector.
+    constexpr std::array<rocsparse_int, n + 1> h_csr_row_ptr = {0, 6, 12, 18, 24, 30, 36};
+
+    // CSR column indices vector.
+    constexpr std::array<rocsparse_int, nnz> h_csr_col_ind
+        = {0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5,
+           0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5};
+
+    // 2. Allocate device memory and offload input data to the device.
+    rocsparse_int* d_csr_row_ptr{};
+    rocsparse_int* d_csr_col_ind{};
+    double*        d_csr_val{};
+    double*        d_LU{};
+    double*        d_data{};
+
+    constexpr size_t size_csr_row_ptr = sizeof(*d_csr_row_ptr) * (n + 1);
+    constexpr size_t size_csr_col_ind = sizeof(*d_csr_col_ind) * nnz;
+    constexpr size_t size_csr_val     = sizeof(*d_csr_val) * nnz;
+    constexpr size_t length_LU        = n * n;
+    constexpr size_t size_LU          = sizeof(*d_LU) * length_LU;
+    rocsparse_int    max_iter         = 100; /*maximum number of iterations*/
+    const size_t     length_data      = 2 * max_iter;
+    const size_t     size_data        = sizeof(*d_data) * length_data;
+
+    HIP_CHECK(hipMalloc(&d_csr_row_ptr, size_csr_row_ptr));
+    HIP_CHECK(hipMalloc(&d_csr_col_ind, size_csr_col_ind));
+    HIP_CHECK(hipMalloc(&d_csr_val, size_csr_val));
+    HIP_CHECK(hipMalloc(&d_LU, size_LU));
+    HIP_CHECK(hipMalloc(&d_data, size_data));
+
+    HIP_CHECK(
+        hipMemcpy(d_csr_row_ptr, h_csr_row_ptr.data(), size_csr_row_ptr, hipMemcpyHostToDevice));
+    HIP_CHECK(
+        hipMemcpy(d_csr_col_ind, h_csr_col_ind.data(), size_csr_col_ind, hipMemcpyHostToDevice));
+    HIP_CHECK(hipMemcpy(d_csr_val, h_csr_val.data(), size_csr_val, hipMemcpyHostToDevice));
+
+    // 3. Initialize rocSPARSE and prepare utility variables for csritilu0 invocation.
+    // Initialize rocSPARSE by creating a handle.
+    rocsparse_handle handle;
+    ROCSPARSE_CHECK(rocsparse_create_handle(&handle));
+
+    // Matrix descriptor.
+    rocsparse_mat_descr descr;
+    ROCSPARSE_CHECK(rocsparse_create_mat_descr(&descr));
+
+    // Iterative ILU0 (incomplete LU factorization with 0 fill-ins) algorithm used.
+    constexpr rocsparse_itilu0_alg alg = rocsparse_itilu0_alg_default;
+
+    // Options selected for the iterative algorithm:
```

Comment:
we should not do this but use the | operator.
option = rocsparse_itilu0_option_stopping_criteria   |
    rocsparse_itilu0_option_compute_nrm_correction |
    rocsparse_itilu0_option_compute_nrm_residual   |
    rocsparse_itilu0_option_convergence_history;

Reply:
Fixed in [5d20e4a](https://github.com/amd/rocm-examples/pull/40/commits/5d20e4a0c2c8b3d2de363b8b865353bab509e6b5)